pallets/cash/src/internal/notices.rs

```diff
@@ -4,13 +4,174 @@ use frame_support::storage::{
 use frame_system::offchain::SubmitTransaction;
 
 use crate::{
-    chains::{ChainAccount, ChainHash, ChainId, ChainSignature, ChainSignatureList},
+    chains::{ChainAccount, ChainAsset, ChainHash, ChainId, ChainSignature, ChainSignatureList},
     log,
-    notices::{has_signer, EncodeNotice, NoticeId, NoticeState},
-    reason::Reason,
-    require, Call, Config, NoticeHashes, NoticeStates, Notices, Validators,
+    notices::{
+        has_signer, CashExtractionNotice, ChangeAuthorityNotice, EncodeNotice, ExtractionNotice,
+        FutureYieldNotice, Notice, NoticeId, NoticeState, SetSupplyCapNotice,
+    },
+    require,
+    types::{
+        AssetAmount, AssetQuantity, CashIndex, CashPrincipalAmount, Reason, Timestamp,
+        ValidatorKeys, APR,
+    },
+    AccountNotices, Call, Config, Event, LatestNotice, Module, NoticeHashes, NoticeHolds,
+    NoticeStates, Notices, Validators,
 };
 
+pub fn dispatch_extraction_notice<T: Config>(
+    asset: ChainAsset,
+    recipient: ChainAccount,
+    amount: AssetQuantity,
+) {
+    dispatch_notice::<T>(
+        recipient.chain_id(),
+        Some(recipient),
+        false,
+        &|notice_id, parent_hash| {
+            Notice::ExtractionNotice(match (asset, recipient, parent_hash) {
+                (
+                    ChainAsset::Eth(eth_asset),
+                    ChainAccount::Eth(eth_account),
+                    ChainHash::Eth(eth_parent_hash),
+                ) => ExtractionNotice::Eth {
+                    id: notice_id,
+                    parent: eth_parent_hash,
+                    asset: eth_asset,
+                    account: eth_account,
+                    amount: amount.value,
+                },
+
+                _ => panic!("XXX not implemented"), // generate these w/ macros?
+            })
+        },
+    )
+}
+
+pub fn dispatch_cash_extraction_notice<T: Config>(
+    recipient: ChainAccount,
+    principal: CashPrincipalAmount,
+) {
+    dispatch_notice::<T>(
+        recipient.chain_id(),
+        Some(recipient),
+        false,
+        &|notice_id, parent_hash| {
+            Notice::CashExtractionNotice(match (recipient, parent_hash) {
+                (ChainAccount::Eth(eth_account), ChainHash::Eth(eth_parent_hash)) => {
+                    CashExtractionNotice::Eth {
+                        id: notice_id,
+                        parent: eth_parent_hash,
+                        account: eth_account,
+                        principal: principal.0,
+                    }
+                }
+
+                _ => panic!("XXX not implemented"), // generate these w/ macros?
```

Comment:
So just checking, we basically just took these from failable into panicing? I get that we can probably reason this, but it's theoretically possible that the ADTs don't match, right? I was hoping we could move the checks into the inputs somehow to make it strictly impossible or the have failed prior to being in "effects" territory.